control-plane/api-gateway/controllers/gatewayclass_controller.go

```diff
@@ -0,0 +1,264 @@
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	gwv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+const (
+	GatewayClassControllerName = "hashicorp.com/consul-api-gateway-controller"
+
+	gatewayClassFinalizer = "gateway-exists-finalizer.consul.hashicorp.com"
+
+	// GatewayClass status fields.
+	invalidParameters = "InvalidParameters"
+	accepted          = "Accepted"
+
+	// GatewayClass status condition reasons.
+	configurationAccepted = "ConfigurationAccepted"
+	configurationInvalid  = "ConfigurationInvalid"
+)
+
+// GatewayClassController reconciles a GatewayClass object.
+// The GatewayClass is responsible for defining the behavior of API gateways
+// which reference the given class.
+type GatewayClassController struct {
+	ControllerName string
+	Log            logr.Logger
+
+	client.Client
+}
+
+// Reconcile handles the reconciliation loop for GatewayClass objects.
+func (r *GatewayClassController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("gatewayClass", req.NamespacedName)
+	log.Info("Reconciling GatewayClass", req.NamespacedName)
+
+	gc := &gwv1beta1.GatewayClass{}
+
+	err := r.Client.Get(ctx, req.NamespacedName, gc)
+	if err != nil {
+		log.Error(err, "unable to get GatewayClass")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if string(gc.Spec.ControllerName) != r.ControllerName {
+		// This GatewayClass is not for this controller.
+		_, err := RemoveFinalizer(ctx, r.Client, gc, gatewayClassFinalizer)
+		if err != nil {
+			log.Error(err, "unable to remove finalizer")
+		}
+
+		return ctrl.Result{}, err
+	}
+
+	if !gc.ObjectMeta.DeletionTimestamp.IsZero() {
+		// We have a deletion request. Ensure we are not in use.
+		used, err := r.isGatewayClassInUse(ctx, gc)
+		if err != nil {
+			log.Error(err, "unable to check if GatewayClass is in use")
+			return ctrl.Result{}, err
+		}
+		if used {
+			log.Info("GatewayClass is in use, cannot delete")
+			return ctrl.Result{}, nil
+		}
+		// Remove our finalizer.
+		if _, err := RemoveFinalizer(ctx, r.Client, gc, gatewayClassFinalizer); err != nil {
+			log.Error(err, "unable to remove finalizer")
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// We are creating or updating the GatewayClass.
+	didUpdate, err := EnsureFinalizer(ctx, r.Client, gc, gatewayClassFinalizer)
+	if err != nil {
+		log.Error(err, "unable to add finalizer")
+		return ctrl.Result{}, err
+	}
+	if didUpdate {
+		// We updated the GatewayClass, requeue to avoid another update.
+		return ctrl.Result{}, nil
+	}
+
+	didUpdate, err = r.validateParametersRef(ctx, gc, log)
+	if didUpdate {
+		if err := r.Client.Update(ctx, gc); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, err
+	}
+	if err != nil {
+		log.Error(err, "unable to validate ParametersRef")
+		return ctrl.Result{}, err
+	}
```

Comment:
So, I think we might be able to write a test for this, but tracing through the logic since this is an idempotent call, assuming we're reconciling something that already has `InvalidParameters` set, then aren't all of these blocks going to be passed and we're going to set the status to `Accepted` as per below? If that's the case I think we'd have an infinite reconciliation loop because we'd be thrashing between `InvalidParameters` and `Accepted`.

Try running a `GatewayClass` with a finalizer already set and a non-existent `GatewayClassConfig` through reconciliation twice and I'm pretty sure you'll see it set `InvalidParameters` and then `Accepted`. If you do this manually I think you'd see reconciliation triggered without termination.